File: agent/sessions/store.py
"""Session registry stored as ``sessions/index.json``.

Each session entry: ``{id, title, created_at, updated_at, preview, version}``.
"""
from __future__ import annotations

import json
import shutil
import uuid
from datetime import datetime, timedelta, timezone
from pathlib import Path

_UTC8 = timezone(timedelta(hours=8))
_VERSION = 1


class SessionStore:
    def __init__(self, root: Path) -> None:
        self.sessions_dir = root / "sessions"
        self.index_path = self.sessions_dir / "index.json"

    # ── internal helpers ────────────────────────────────────────────
    def _load(self) -> list[dict]:
        if not self.index_path.exists():
            return []
        try:
            raw = self.index_path.read_text(encoding="utf-8").strip()
            if not raw:
                return []
            data = json.loads(raw)
            if not isinstance(data, list):
                raise ValueError("index.json must be a list")
            changed = False
            normalized = []
            for item in data:
                if not isinstance(item, dict):
                    changed = True
                    continue
                clean = self._normalize(item)
                changed = changed or clean != item
                normalized.append(clean)
            if changed:
                self._save(normalized)
            return normalized
        except (json.JSONDecodeError, ValueError, OSError):
            # Quarantine corrupt index, start fresh
            ts = datetime.now(_UTC8).strftime("%Y%m%dT%H%M%S")
            corrupt_path = self.sessions_dir / f"index.corrupt-{ts}.json"
            try:
                shutil.move(str(self.index_path), str(corrupt_path))
            except OSError:
                pass
            return []

    def _save(self, items: list[dict]) -> None:
        self.sessions_dir.mkdir(parents=True, exist_ok=True)
        tmp = self.index_path.with_suffix(".json.tmp")
        tmp.write_text(
            json.dumps(items, ensure_ascii=False, indent=2) + "\n",
            encoding="utf-8",
        )
        tmp.replace(self.index_path)

    def _stamp(self) -> str:
        return datetime.now(_UTC8).strftime("%Y-%m-%dT%H:%M:%S%z")

    def _dir(self, session_id: str) -> Path:
        return self.sessions_dir / session_id

    def session_dir(self, session_id: str) -> Path:
        return self._dir(session_id)

    # ── public API ──────────────────────────────────────────────────
    def list(self, *, include_archived: bool = False) -> list[dict]:
        items = self._load()
        if not include_archived:
            items = [item for item in items if not item.get("archived_at")]
        items.sort(key=lambda x: x.get("updated_at", ""), reverse=True)
        return items

    def create(
        self,
        title: str = "",
        *,
        title_status: str | None = None,
        mode: str = "chat",
        project: dict | None = None,
    ) -> dict:
        now = self._stamp()
        sid = uuid.uuid4().hex[:16]
        clean_title = title.strip()
        clean_mode = "build" if mode == "build" else "chat"
        project = project or {}
        entry = {
            "id": sid,
            "title": clean_title or "Untitled",
            "created_at": now,
            "updated_at": now,
            "preview": "",
            "message_count": 0,
            "title_status": title_status or ("manual" if clean_title else "placeholder"),
            "mode": clean_mode,
            "project_id": str(project.get("project_id") or "") or None,
            "project_path": str(project.get("project_path") or "") or None,
            "project_name": str(project.get("project_name") or "") or None,
            "archived_at": None,
            "version": _VERSION,
        }
        items = self._load()
        items.append(entry)
        self._save(items)
        self._dir(sid).mkdir(parents=True, exist_ok=True)
        return entry

    def get(self, session_id: str) -> dict | None:
        for item in self._load():
            if item.get("id") == session_id:
                return item
        return None

    def delete(self, session_id: str) -> bool:
        items = self._load()
        if len(items) <= 1:
            return False
        idx = next((i for i, e in enumerate(items) if e.get("id") == session_id), None)
        if idx is None:
            return False
        del items[idx]
        self._save(items)
        shutil.rmtree(self._dir(session_id), ignore_errors=True)
        return True

    def rename(self, session_id: str, title: str) -> bool:
        items = self._load()
        for e in items:
            if e.get("id") == session_id:
                e["title"] = title.strip()
                e["updated_at"] = self._stamp()
                e["title_status"] = "manual"
                self._save(items)
                return True
        return False

    def archive(self, session_id: str) -> dict | None:
        return self._set_archived(session_id, archived=True)

    def restore(self, session_id: str) -> dict | None:
        return self._set_archived(session_id, archived=False)

    def set_generated_title(self, session_id: str, title: str) -> dict | None:
        items = self._load()
        for e in items:
            if e.get("id") == session_id:
                e["title"] = title.strip()
                e["title_status"] = "generated"
                e["updated_at"] = self._stamp()
                self._save(items)
                return dict(e)
        return None

    def touch(self, session_id: str, preview: str, *, increment_messages: bool = False) -> dict | None:
        items = self._load()
        for e in items:
            if e.get("id") == session_id:
                e["preview"] = preview[:280]
                if increment_messages:
                    e["message_count"] = int(e.get("message_count") or 0) + 1
                e["updated_at"] = self._stamp()
                self._save(items)
                return dict(e)
        return None

    def _normalize(self, item: dict) -> dict:
        clean = dict(item)
        mode = str(clean.get("mode") or "chat").strip().lower()
        if mode != "build":
            mode = "chat"
        clean["mode"] = mode
        clean.setdefault("message_count", 0)
        clean.setdefault("title_status", "manual")
        clean["project_id"] = str(clean.get("project_id") or "") or None
        clean["project_path"] = str(clean.get("project_path") or "") or None
        clean["project_name"] = str(clean.get("project_name") or "") or None
        clean["archived_at"] = str(clean.get("archived_at") or "") or None
        clean["version"] = int(clean.get("version") or _VERSION)
        return clean

    def _set_archived(self, session_id: str, *, archived: bool) -> dict | None:
        items = self._load()
        for e in items:
            if e.get("id") == session_id:
                e["archived_at"] = self._stamp() if archived else None
                e["updated_at"] = self._stamp()
                self._save(items)
                return dict(e)
        return None

File: packages/core/src/sessions/store.ts
import { randomUUID } from 'node:crypto'
import { existsSync, mkdirSync, readFileSync, renameSync, rmSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'

const VERSION = 1

export interface SessionEntry {
  id: string
  title: string
  created_at: string
  updated_at: string
  preview: string
  message_count: number
  title_status: string
  mode: 'chat' | 'build'
  project_id: string | null
  project_path: string | null
  project_name: string | null
  archived_at: string | null
  version: number
}

export interface SessionCreateOptions {
  titleStatus?: string | null
  mode?: string
  project?: Record<string, unknown> | null
}

export class SessionStore {
  readonly root: string
  readonly sessionsDir: string
  readonly indexPath: string

  constructor(root: string) {
    this.root = root
    this.sessionsDir = join(root, 'sessions')
    this.indexPath = join(this.sessionsDir, 'index.json')
  }

  sessionDir(sessionId: string): string {
    return join(this.sessionsDir, sessionId)
  }

  list(opts: { includeArchived?: boolean } = {}): SessionEntry[] {
    let items = this.load()
    if (!opts.includeArchived) items = items.filter((item) => !item.archived_at)
    items.sort((a, b) => String(b.updated_at ?? '').localeCompare(String(a.updated_at ?? '')))
    return items
  }

  create(title = '', opts: SessionCreateOptions = {}): SessionEntry {
    const now = stamp()
    const cleanTitle = title.trim()
    const mode = opts.mode === 'build' ? 'build' : 'chat'
    const project = opts.project ?? {}
    const entry: SessionEntry = {
      id: randomUUID().replace(/-/g, '').slice(0, 16),
      title: cleanTitle || 'Untitled',
      created_at: now,
      updated_at: now,
      preview: '',
      message_count: 0,
      title_status: opts.titleStatus || (cleanTitle ? 'manual' : 'placeholder'),
      mode,
      project_id: nullableText(project.project_id),
      project_path: nullableText(project.project_path),
      project_name: nullableText(project.project_name),
      archived_at: null,
      version: VERSION,
    }
    const items = this.load()
    items.push(entry)
    this.save(items)
    mkdirSync(this.sessionDir(entry.id), { recursive: true })
    return entry
  }

  get(sessionId: string): SessionEntry | null {
    return this.load().find((item) => item.id === sessionId) ?? null
  }

  delete(sessionId: string): boolean {
    const items = this.load()
    if (items.length <= 1) return false
    const idx = items.findIndex((item) => item.id === sessionId)
    if (idx < 0) return false
    items.splice(idx, 1)
    this.save(items)
    rmSync(this.sessionDir(sessionId), { recursive: true, force: true })
    return true
  }

  rename(sessionId: string, title: string): boolean {
    const items = this.load()
    for (const item of items) {
      if (item.id !== sessionId) continue
      item.title = title.trim()
      item.updated_at = stamp()
      item.title_status = 'manual'
      this.save(items)
      return true
    }
    return false
  }

  archive(sessionId: string): SessionEntry | null {
    return this.setArchived(sessionId, true)
  }

  restore(sessionId: string): SessionEntry | null {
    return this.setArchived(sessionId, false)
  }

  setGeneratedTitle(sessionId: string, title: string): SessionEntry | null {
    const items = this.load()
    for (const item of items) {
      if (item.id !== sessionId) continue
      item.title = title.trim()
      item.title_status = 'generated'
      item.updated_at = stamp()
      this.save(items)
      return { ...item }
    }
    return null
  }

  touch(sessionId: string, preview: string, opts: { incrementMessages?: boolean } = {}): SessionEntry | null {
    const items = this.load()
    for (const item of items) {
      if (item.id !== sessionId) continue
      item.preview = preview.slice(0, 280)
      if (opts.incrementMessages) item.message_count = Number(item.message_count || 0) + 1
      item.updated_at = stamp()
      this.save(items)
      return { ...item }
    }
    return null
  }

  private load(): SessionEntry[] {
    if (!existsSync(this.indexPath)) return []
    try {
      const text = readFileSync(this.indexPath, 'utf8').trim()
      if (!text) return []
      const data = JSON.parse(text)
      if (!Array.isArray(data)) throw new Error('index.json must be a list')
      let changed = false
      const normalized: SessionEntry[] = []
      for (const item of data) {
        if (!item || typeof item !== 'object' || Array.isArray(item)) {
          changed = true
          continue
        }
        const clean = normalizeSession(item as Record<string, unknown>)
        if (JSON.stringify(clean) !== JSON.stringify(item)) changed = true
        normalized.push(clean)
      }
      if (changed) this.save(normalized)
      return normalized
    } catch {
      this.quarantineIndex()
      return []
    }
  }

  private save(items: SessionEntry[]): void {
    mkdirSync(this.sessionsDir, { recursive: true })
    const tmp = this.indexPath.replace(/\.json$/, '.json.tmp')
    writeFileSync(tmp, JSON.stringify(items, null, 2) + '\n', 'utf8')
    renameSync(tmp, this.indexPath)
  }

  private quarantineIndex(): void {
    mkdirSync(this.sessionsDir, { recursive: true })
    if (!existsSync(this.indexPath)) return
    const target = join(this.sessionsDir, `index.corrupt-${stampForFilename()}.json`)
    try { renameSync(this.indexPath, target) } catch { /* ignore */ }
  }

  private setArchived(sessionId: string, archived: boolean): SessionEntry | null {
    const items = this.load()
    for (const item of items) {
      if (item.id !== sessionId) continue
      item.archived_at = archived ? stamp() : null
      item.updated_at = stamp()
      this.save(items)
      return { ...item }
    }
    return null
  }
}

function normalizeSession(raw: Record<string, unknown>): SessionEntry {
  const mode = String(raw.mode ?? 'chat').trim().toLowerCase() === 'build' ? 'build' : 'chat'
  return {
    id: String(raw.id ?? ''),
    title: String(raw.title ?? 'Untitled'),
    created_at: String(raw.created_at ?? raw.createdAt ?? ''),
    updated_at: String(raw.updated_at ?? raw.updatedAt ?? ''),
    preview: String(raw.preview ?? ''),
    message_count: toInt(raw.message_count ?? raw.messageCount, 0),
    title_status: String(raw.title_status ?? raw.titleStatus ?? 'manual'),
    mode,
    project_id: nullableText(raw.project_id ?? raw.projectId),
    project_path: nullableText(raw.project_path ?? raw.projectPath),
    project_name: nullableText(raw.project_name ?? raw.projectName),
    archived_at: nullableText(raw.archived_at ?? raw.archivedAt),
    version: toInt(raw.version, VERSION),
  }
}

function nullableText(value: unknown): string | null {
  const text = String(value ?? '').trim()
  return text || null
}

function toInt(value: unknown, fallback: number): number {
  const n = typeof value === 'number' ? Math.trunc(value) : Number.parseInt(String(value), 10)
  return Number.isFinite(n) ? n : fallback
}

function stamp(): string {
  const d = new Date(Date.now() + 8 * 3600 * 1000)
  const pad = (n: number) => String(n).padStart(2, '0')
  return `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}T${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}+0800`
}

function stampForFilename(): string {
  return stamp().replace(/[-:]/g, '').replace('+0800', '')
}
